### PR TITLE
Add support for system-level `uv.toml` configuration

### PR DESCRIPTION
**Pull request by amjith**
## Summary

Look for a system level uv.toml config file under `/etc/uv/uv.toml` or `C:\ProgramData`. 

This PR is to address #6742 and start a conversation. 

## Test Plan

This was tested locally manually on MacOS. I am happy to contribute tests once we settle on the approach. 

cc @thatch

### CODE REVIEW

**Comment by zanieb** [comment]
Why not use the XDG defaults?

I'm not sure if we'll need to address https://github.com/astral-sh/uv/issues/6742#issuecomment-2383897018

**Review by charliermarsh — APPROVED** [review]
The code looks good to me, assuming we're happy with the paths! I'll defer to @zanieb on the choice of path.

**Comment by charliermarsh** [comment]
\cc @pradyunsg

**Comment by amjith** [comment]
> Why not use the XDG defaults?

From reading the XDG docs it seemed like `/etc/xdg/uv/uv.toml` is the path in linux and MacOS doesn't really have an equivalent but we could use `~/Library/Preferences /Library/Application Support /Library/Preferences`. 

I decided to go with a stable location that is shorter and the same between Linux and MacOS. I'm happy to check for the env var and use that if it exists and fallback to `/etc/uv/uv.toml` if it doesn't exist. 

Let me know your preference and I can adjust this accordingly. 

Testing:
I'm new to rust hence not too familiar with the testing story. Are there relevant tests I should update or add?

**Comment by charliermarsh** [comment]
> Why not use the XDG defaults?

What _are_ the XDG defaults for this?

**Comment by zanieb** [comment]
> $XDG_CONFIG_DIRS defines the preference-ordered set of base directories to search for configuration files in addition to the $XDG_CONFIG_HOME base directory. The directories in $XDG_CONFIG_DIRS should be separated with a colon ':'.
>
> If $XDG_CONFIG_DIRS is either not set or empty, a value equal to `/etc/xdg` should be used.

**Comment by zanieb** [comment]
Basically, at the very least, we need to respect `XDG_CONFIG_DIRS` and look in all of the directories given there. If not set, we should also look in `/etc/xdg`. I'm down to read `/etc/uv` in addition to all of the XDG paths, but we need to decide if it's lower or higher priority than the XDG ones. I think we should do the same thing on macOS as we do on Linux.

**Comment by charliermarsh** [comment]
Okay cool, makes sense. How's this for a proposal:

1. If `XDG_CONFIG_DIRS` is defined and non-empty, check each `${DIR}/uv/uv.toml`.
2. If it's not defined or empty, check `/etc/xdg/uv/uv.toml`.
3. Check `/etc/uv/uv.toml`.

For rationale: we store in `/etc/uv/uv.toml` rather than `/etc/uv.toml` based on ([ref](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html))...

> It is recommended that files be stored in subdirectories of /etc rather than directly in /etc.

Additionally, we check `/etc/uv/uv.toml` _last_ (rather than first) so that users can override configuration by adding to `XDG_CONFIG_DIRS`, rather than `/etc/uv/uv.toml` always taking precedent (rare, but we have to decide one way or the other).

---

How does this sound? @amjith -- are you interested in implementing these changes?

**Comment by amjith** [comment]
Yes. I'm happy to do that. I'll take a stab at it this Friday.

**Comment by zanieb** [comment]
And we combine each config we find?

**Comment by amjith** [comment]
Should it combine them all or should it stop at the first one it finds?

**Comment by amjith** [comment]
The current behavior is to combine project + system + user. If we choose to combine config from `XDG_CONFIG_DIRS` and `/etc/uv/uv.toml` and call that the system config that could get confusing. Especially since `XDG_CONFIG_DIRS` can have multiple folders and each one could have a uv.toml file.

**Comment by charliermarsh** [comment]
I'd probably assume that we only take the first one.

**Comment by zanieb** [comment]
XDG says you can do either :)

> A specification that refers to $XDG_DATA_DIRS or $XDG_CONFIG_DIRS should define what the behaviour must be when a file is located under multiple base directories. It could, for example, define that only the file under the most important base directory should be used or, as another example, it could define rules for merging the information from the different files.

We should probably choose one. Not merging seems simpler, I agree. Are there other tools that respect these that we can refer to though?

File: crates/uv-settings/src/lib.rs
Comment (thatch) [inline crates/uv-settings/src/lib.rs:204]: I went to re-read the spec, which says

> If $XDG_CONFIG_DIRS is either not set or empty, a value equal to /etc/xdg should be used.

I think this means don't fall back to /etc/xdg if the env var contains even just `:`

If that's true you should handle the "/etc/xdg" default here and not in `system_config_file()`.  To make things easier you could pass in the env var (so you don't have to set it in the test), since this is only used internally.

File: crates/uv-settings/src/lib.rs
Comment (thatch) [inline crates/uv-settings/src/lib.rs:63]: You can use the same pattern here as from line 37, to save some indent by putting the None up top.

File: crates/uv-settings/src/lib.rs
Comment (thatch) [inline crates/uv-settings/src/lib.rs:64]: Now that the 3rd case (not a dir) is removed, you can just `let options = read_file(&file)?` I think.

See https://doc.rust-lang.org/beta/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator

File: crates/uv/src/lib.rs
Comment (thatch) [inline crates/uv/src/lib.rs:131]: Can you double-check the order of the `combine` calls?  It looks like `project.combine(user)` is in decreasing order, so I'd expect `system` to come last?

File: crates/uv-settings/src/lib.rs
Comment (thatch) [inline crates/uv-settings/src/lib.rs:272]: fixtureS

**Comment by amjith** [comment]
I'm happy with the code changes in the PR. Let me know your feedback. 

Unfortunately the compilation fails in windows and I could use a hint on how to resolve that.

**Comment by amjith** [comment]
I'm stuck on the Windows test failing. I don't have a windows laptop to try this out and somehow the env vars in Windows aren't working as I assumed it would. 

If you you folks have any suggestions, I'm all ears.

**Comment by charliermarsh** [comment]
No worries — I can take a look tomorrow (when I’m back home) on my Windows machine.

**Review by Dominik50111 — APPROVED** [review]

File: crates/uv-settings/src/lib.rs
Comment (charliermarsh) [inline crates/uv-settings/src/lib.rs:204]: Does `env::split_paths` work here?

Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:204]: Empirically, appears not!

**Comment by charliermarsh** [comment]
I think the Windows test is failing because you're expecting a `ProgramData` child that doesn't exist in the fixture. I'm gonna refactor the fixtures slightly to write the files within the tests themselves.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [reply crates/uv-settings/src/lib.rs:204]: Really? That's pretty surprising.

Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:204]: Yeah. It returns `""` in the iterator, which then resolves to the current working directory.

Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:204]: (We could filter empty string I guess.)

**Comment by charliermarsh** [comment]
Should we be using `%ProgramData%` rather than system drive?

**Comment by charliermarsh** [comment]
This looks good to me -- tests are passing, and I added some docs. I'll give @zanieb some time to review too though before merging.

**Comment by charliermarsh** [comment]
I sort of think we should use `SHGetKnownFolderPath(FOLDERID_ProgramData)` instead of the current approach on Windows, but I may do it as a separate change and it's not blocking.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [reply crates/uv-settings/src/lib.rs:204]: Oh interesting. Yeah I guess it seems more correct to handle the empty string separately. I don't feel strongly since it's already platform-specific code though.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [inline crates/uv-settings/src/lib.rs:217]: It seems like this could use a little touchup for readability.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [inline crates/uv-settings/src/lib.rs:291]: Why is this in a block?

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [inline crates/uv-settings/src/lib.rs:304]: This isn't thread-safe and our tests run concurrently.

Unfortunately this isn't easy to solve. In the Python discovery tests, we use this pattern:

https://github.com/astral-sh/uv/blob/01c44af3c3ffc9f5e78ef4b415305350c029d1e0/crates/uv-python/src/lib.rs#L44-L55

https://github.com/astral-sh/uv/blob/01c44af3c3ffc9f5e78ef4b415305350c029d1e0/crates/uv-python/src/tests.rs#L78-L105

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [inline crates/uv-settings/src/lib.rs:348]: Same as https://github.com/astral-sh/uv/pull/7851/files#r1808963874

File: docs/configuration/files.md
Comment (zanieb) [inline docs/configuration/files.md:44]: We should clarify that we'll take the first system-level configuration file we find, not merge them all.

Comment (zanieb) [reply docs/configuration/files.md:44]: Maybe below in the merging paragraph?

File: crates/uv-settings/src/lib.rs
Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:304]: Why do we even need this? The tests pass without changing the directory.

File: crates/uv-settings/src/lib.rs
Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:291]: To clarify the scope.

File: crates/uv-settings/src/lib.rs
Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:348]: This is different than the above. The above is changing the current directory. This is changing an env var. (The same points may apply, I just wanted to point it out.)

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [reply crates/uv-settings/src/lib.rs:304]: I'm not sure. Seems like we don't then? I don't know why we'd read the current directory for this? Maybe if you have `XDG_CONFIG_PATHS=foo::bar` and we parse something relative to the working directory? Seems weird.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [reply crates/uv-settings/src/lib.rs:348]: Yeah thanks for clarifying, the same problem is true unfortunately. If there's another test that reads `SYSTEMDRIVE` we can encounter a race condition and the tests will be flaky.

Comment (charliermarsh) [reply crates/uv-settings/src/lib.rs:348]: Okay cool, got it!

Comment (zanieb) [reply crates/uv-settings/src/lib.rs:348]: Unfortunately I think this may still be a problem even if you use `with_vars` because [temp-env](https://crates.io/crates/temp-env) creates a mutex to prevent concurrent conflicts between variables but if we're not using `temp-env` elsewhere we can still end up mutating their state. This isn't a big problem now because there aren't tests that rely on reading the environment that we're mutating in the Python discovery tests. The same is true of this test now, but it may not be true forever.

File: crates/uv-settings/src/lib.rs
Comment (zanieb) [reply crates/uv-settings/src/lib.rs:291]: The scope of changing the current directory? I think it's misleading because it's not like `set_current_dir` reverts when the scope exits.

**Comment by charliermarsh** [comment]
I'll tweak and merge this today.

**Comment by gaby** [comment]
@charliermarsh Is this in the latest release or the next one?

Thanks for this, looking forward to automating `uv` using ansible

**Comment by charliermarsh** [comment]
It's not released yet, maybe today :)

**Comment by wimglenn** [comment]
sysadmin note - if you arrived here looking for system-level uv config in /etc, the minimum uv version needed is [0.4.26](https://github.com/astral-sh/uv/releases/tag/0.4.26)

**Comment by gaby** [comment]
@wimglenn Great start but its missing all the ENV fields from uv. There's no way to setup a proxy via uv.toml either

**Comment by charliermarsh** [comment]
@gaby -- Can I kindly request that you refrain from reposting that comment across multiple issues / PRs? You've mentioned it in a few places already, and it's best tracked elsewhere. Thanks.

**Pull request merged**
